### DISCUSSION TITLE
Help understanding how to do Key Reuse with different thresholds

### DISCUSSION THREAD

**Discussion by miketwenty1** [rust-bitcoin/rust-miniscript#738] [Q&A]
I know this syntax is incorrect, but I'm unsure about the best way to implement conditional logic branches.
```rust
let pol_str = "or(
    and(
        thresh(3, pk(A), pk(B), pk(C), pk(D)),
        older(1000)
    ),
    and(
        thresh(2, pk(A), pk(B), pk(C), pk(D)),
        older(3000)
    ),
    pk(E)
)"
.replace(&[' ', '\n', '\t'][..], "");
```

What would be a syntax strategy for implementing 2 timelock threshold conditions like this?
```
or(
    and(
        thresh(3, pk(A), pk(B), pk(C), pk(D)),
        older(1000)
    ),
    and(
        thresh(2, pk(A), pk(B), pk(C), pk(D)),
        older(3000)
    )
)
```

**Comment by apoelstra** [comment]
In this case, I would use multiple Tapbranches. There is no problem in principle with key reuse across tapbranches (though I don't think that this library understands that -- our validation logic is very crufty and I've been struggling to organize it lately -- see #723).

If there is a specific key (say, `D`) which you want to stop being necessary after the timeout, you can try rearranging your policy like

```
and(
    thresh(3, pk(A), pk(B), pk(C), or(pk(D), older(3000))),
    older(1000)
)
```

But in the general case you describe, this might not be possible right now in Miniscript. (With the caveat that it *should* be possible with Taproot and I just don't know if this library is smart enough to understand that.)

If Miniscript itself were smarter, you would be able to directly make the threshold value parameterized by timelocks. Given the popularity of these "decaying timelock" constructions this might be something that we should consider adding to the language..

**Reply by miketwenty1** [reply]
I want to make sure I understand the semantics of a tapbranch.

> I would use multiple Tapbranches.

Does that mean ideally you would go with my above approach:
Branch 1:  `and(thresh(3, pk(A), pk(B), pk(C), pk(D)),older(1000))`
Branch 2: `and(thresh(2, pk(A), pk(B), pk(C), pk(D)),older(3000))`

Or are you saying your example is multiple tapbranches using a single `and`?
> and(
    thresh(3, pk(A), pk(B), pk(C), or(pk(D), older(3000))),
    older(1000)
)

**Reply by apoelstra** [reply]
>Does that mean ideally you would go with my above approach:

Yep, exactly.
